src/devices/machine/i8291a.h

```diff
@@ -0,0 +1,415 @@
+
+#define MCFG_I8291A_DIO_WRITE_CB(_write)                   \
+	downcast<i8291a_device &>(*device).set_dio_write_cb(DEVCB_##_write);
+
+#define MCFG_I8291A_EOI_WRITE_CB(_write)                               \
+	downcast<i8291a_device &>(*device).set_eoi_write_cb(DEVCB_##_write);
+
+#define MCFG_I8291A_DAV_WRITE_CB(_write)                               \
+	downcast<i8291a_device &>(*device).set_dav_write_cb(DEVCB_##_write);
+
+#define MCFG_I8291A_NRFD_WRITE_CB(_write)                              \
+	downcast<i8291a_device &>(*device).set_nrfd_write_cb(DEVCB_##_write);
+
+#define MCFG_I8291A_NDAC_WRITE_CB(_write)                              \
+	downcast<i8291a_device &>(*device).set_ndac_write_cb(DEVCB_##_write);
+
+#define MCFG_I8291A_SRQ_WRITE_CB(_write)                               \
+	downcast<i8291a_device &>(*device).set_srq_write_cb(DEVCB_##_write);
+
+
+class i8291a_device : public device_t
+{
+public:
+	// construction/destruction
+	i8291a_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	template <class Object> devcb_base& set_int_write_cb(Object &&cb)
+		{ return m_int_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_dreq_write_cb(Object &&cb)
+		{ return m_dreq_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_trig_write_cb(Object &&cb)
+		{ return m_trig_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_eoi_write_cb(Object &&cb)
+		{ return m_eoi_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_dav_write_cb(Object &&cb)
+		{ return m_dav_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_nrfd_write_cb(Object &&cb)
+		{ return m_nrfd_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_ndac_write_cb(Object &&cb)
+		{ return m_ndac_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_srq_write_cb(Object &&cb)
+		{ return m_srq_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_dio_write_cb(Object &&cb)
+		{ return m_dio_write_func.set_callback(std::forward<Object>(cb)); }
+
+	template <class Object> devcb_base& set_dio_read_cb(Object &&cb)
+		{ return m_dio_read_func.set_callback(std::forward<Object>(cb)); }
+
+	// Signal inputs
+	DECLARE_WRITE_LINE_MEMBER(reset_w);
+	DECLARE_WRITE_LINE_MEMBER(dack_w);
+
+	// signal output
+	devcb_write_line m_int_write_func;
+	devcb_write_line m_dreq_write_func;
+	devcb_write_line m_trig_write_func;
+	devcb_write_line m_eoi_write_func;
+	devcb_write_line m_dav_write_func;
+	devcb_write_line m_nrfd_write_func;
+	devcb_write_line m_ndac_write_func;
+	devcb_write_line m_srq_write_func;
+
+	devcb_write8 m_dio_write_func;
+	devcb_read8 m_dio_read_func;
```

Comment:
Please don't expose internal stuff like this - if you need it to be public, there's probably something wrong with your design.